runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -2353,3 +2354,186 @@ J9::TransformUtil::specializeInvokeExactSymbol(TR::Compilation *comp, TR::Node *
+   auto symRef = node->getSymbolReference();
+   // Refine the call
+   auto refinedMethod = fej9->createResolvedMethod(comp->trMemory(), targetMethod, symRef->getOwningMethod(comp));
+   if (!performTransformation(comp, "O^O Refine invokeBasic n%dn %p with known MH object\n", node->getGlobalIndex(), node))
+      return false;
+
+   TR_ResolvedJ9Method::setAdapterOrLambdaForm(refinedMethod);
+
+   // Preserve NULLCHK
+   TR::TransformUtil::separateNullCheck(comp, treetop, trace);
+
+   TR::SymbolReference *newSymRef =
+       comp->getSymRefTab()->findOrCreateMethodSymbol
+       (symRef->getOwningMethodIndex(), -1, refinedMethod, TR::MethodSymbol::Static);
+
+   TR::Node::recreateWithSymRef(node, refinedMethod->directCallOpCode(), newSymRef);
+   return true;
+#else
+   return false;
+#endif
+   }
+
+TR::MethodSymbol::Kinds getTargetMethodCallKind(TR::RecognizedMethod rm)
+   {
+   TR::MethodSymbol::Kinds callKind;
+   switch (rm)
+      {
+      case TR::java_lang_invoke_MethodHandle_invokeBasic:
+      case TR::java_lang_invoke_MethodHandle_linkToStatic:
+         callKind = TR::MethodSymbol::Static; break;
+      case TR::java_lang_invoke_MethodHandle_linkToSpecial:
+         callKind = TR::MethodSymbol::Special; break;
+      case TR::java_lang_invoke_MethodHandle_linkToVirtual:
+         callKind = TR::MethodSymbol::Virtual; break;
+      case TR::java_lang_invoke_MethodHandle_linkToInterface:
+         callKind = TR::MethodSymbol::Interface; break;
+      default:
+         TR_ASSERT(0, "Unsupported method");
+      }
+   return callKind;
+   }
+
+// Use getIndirectCall(datatype), pass in return type
+TR::ILOpCodes getTargetMethodCallOpCode(TR::RecognizedMethod rm, TR::DataType type)
+   {
+   switch (rm)
+      {
+      case TR::java_lang_invoke_MethodHandle_invokeBasic:
+      case TR::java_lang_invoke_MethodHandle_linkToStatic:
+      case TR::java_lang_invoke_MethodHandle_linkToSpecial:
+         return TR::ILOpCode::getDirectCall(type);
+      case TR::java_lang_invoke_MethodHandle_linkToVirtual:
+      case TR::java_lang_invoke_MethodHandle_linkToInterface:
+         return TR::ILOpCode::getIndirectCall(type);
+      default:
+         TR_ASSERT(0, "Unsupported method");
+      }
+   return TR::BadILOp;
+   }
+
+bool
+J9::TransformUtil::refineMethodHandleLinkTo(TR::Compilation* comp, TR::TreeTop* treetop, TR::Node* node, TR::KnownObjectTable::Index mnIndex, bool trace)
+   {
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+   auto symRef = node->getSymbolReference();
+   auto rm = node->getSymbol()->castToMethodSymbol()->getMandatoryRecognizedMethod();
+   switch (rm)
+      {
+      case TR::java_lang_invoke_MethodHandle_linkToStatic:
+      case TR::java_lang_invoke_MethodHandle_linkToSpecial:
+      case TR::java_lang_invoke_MethodHandle_linkToVirtual:
+         break;
+      default:
+        TR_ASSERT(false, "Unsupported method %s", symRef->getSymbol()->getResolvedMethodSymbol()->getResolvedMethod()->signature(comp->trMemory()));
```

Comment:
Since this line should never be reached, we could use TR_ASSERT_FATAL without the fear of extra overhead

Reply:
A reminder to change this into an ASSERT_FATAL

Reply:
Fixed.